### PR TITLE
Prevent cache write from exceeding IOV_MAX

### PR DESCRIPTION
**Pull request by zacw7**
Differential Revision: D55941557

### CODE REVIEW

**Comment by amitkdutta** [comment]
This is a very long lasting bug. Thanks @zacw7 for working on it.
CC: @czentgr @majetideepak

File: velox/common/caching/SsdFile.cpp [hunk 1]
```diff
@@ -81,6 +81,13 @@ void addEntryToIovecs(AsyncDataCacheEntry& entry, std::vector<iovec>& iovecs) {
     };
   }
 }
+
+int32_t numIoVectorsFromEntry(AsyncDataCacheEntry& entry) {
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:85]:
```
// Returns the number of entries in a cache 'entry'.
```

File: velox/common/caching/SsdFile.cpp [hunk 2]
```diff
@@ -438,6 +450,26 @@ void SsdFile::write(std::vector<CachePin>& pins) {
   }
 }
 
+bool SsdFile::write(
+    std::vector<iovec>& iovecs,
+    uint64_t offset,
+    uint64_t expectedWrittenBytes) {
+  const auto rc = folly::pwritev(fd_, iovecs.data(), iovecs.size(), offset);
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:457]:
```
nit: s/rc/ret/
```

File: velox/common/caching/SsdFile.cpp [hunk 3]
```diff
@@ -376,40 +383,45 @@ void SsdFile::write(std::vector<CachePin>& pins) {
...
     int32_t numWritten = 0;
-    int32_t bytes = 0;
+    int32_t bytesToWrite = 0;
     std::vector<iovec> iovecs;
+    uint64_t writeOffset = offset;
     for (auto i = storeIndex; i < pins.size(); ++i) {
       auto* entry = pins[i].checkedEntry();
       const auto entrySize = entry->size();
-      if (bytes + entrySize > available) {
+      const auto numOfEntryIoVecs = numIoVectorsFromEntry(*entry);
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:400]:
```
nit: s/numOfEntryIoVecs/numIovecs/
```

File: velox/common/caching/SsdFile.cpp [hunk 4]
```diff
@@ -376,40 +383,45 @@ void SsdFile::write(std::vector<CachePin>& pins) {
...
+      VELOX_CHECK_LE(numOfEntryIoVecs, IOV_MAX);
+      if (iovecs.size() + numOfEntryIoVecs > IOV_MAX) {
+        // Flush first if adding the current entry will make iovecs size exceed
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:403]:
```
// Writes out the accumulated iovecs if it exceeds IOV_MAX limit.
```

File: velox/common/caching/SsdFile.h [hunk 5]
```diff
@@ -328,6 +328,13 @@ class SsdFile {
   // the files for making new checkpoints.
   void initializeCheckpoint();
 
+  // Writes 'iovecs' to the SSD file at the 'offset'. Returns true if the write
+  // succeeds; otherwise, log the error and return false.
+  bool write(
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.h:333]:
```
bool write(uint64_t offset, uint64_t length, std::vector<iovec>& iovecs)
```

File: velox/common/caching/SsdFile.cpp [hunk 6]
```diff
@@ -438,6 +450,26 @@ void SsdFile::write(std::vector<CachePin>& pins) {
...
+        << "Failed to write to SSD, file name: " << fileName_ << ", fd: " << fd_
+        << ", size: " << iovecs.size() << ", offset: " << offset
+        << ", error code: " << errno
+        << ", error string: " << folly::errnoStr(errno);
+    ++stats_.writeSsdErrors;
+    // If write fails, we return without adding the pins to the cache. The
+    // entries are unchanged.
+    return false;
+  }
+  iovecs.clear();
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:469]: Let the caller to clear iovecs and then pass const reference (const std::vector<iovec>& iovecs) to write function?

File: velox/common/caching/tests/SsdFileTest.cpp [hunk 7]
```diff
@@ -302,6 +302,15 @@ TEST_F(SsdFileTest, writeAndRead) {
       }
     }
   }
+
+  // Test cache writes with different iobufs sizes.
+  for (int numPins : {0, 1, IOV_MAX - 1, IOV_MAX, IOV_MAX + 1}) {
```
Comment (xiaoxmeng) [inline velox/common/caching/tests/SsdFileTest.cpp:307]:
```
SCOPED_TRACE(fmt::format("numPins: {}", numPins));
```

File: velox/common/caching/SsdFile.cpp [hunk 8]
```diff
@@ -376,40 +383,45 @@ void SsdFile::write(std::vector<CachePin>& pins) {
...
-      if (bytes + entrySize > available) {
+      const auto numOfEntryIoVecs = numIoVectorsFromEntry(*entry);
+      VELOX_CHECK_LE(numOfEntryIoVecs, IOV_MAX);
+      if (iovecs.size() + numOfEntryIoVecs > IOV_MAX) {
+        // Flush first if adding the current entry will make iovecs size exceed
+        // IOV_MAX.
+        if (!write(iovecs, writeOffset, bytesToWrite)) {
+          return;
+        }
+        writeOffset += bytesToWrite;
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:408]:
```
iovecs.clear();
writeOffset += bytesToWrite;
...
```

File: velox/common/caching/SsdFile.cpp [hunk 9]
```diff
@@ -376,40 +383,45 @@ void SsdFile::write(std::vector<CachePin>& pins) {
...
-      VELOX_SSD_CACHE_LOG(ERROR)
-          << "Failed to write to SSD, file name: " << fileName_
-          << ", fd: " << fd_ << ", size: " << iovecs.size()
-          << ", offset: " << offset << ", error code: " << errno
-          << ", error string: " << folly::errnoStr(errno);
-      ++stats_.writeSsdErrors;
-      // If write fails, we return without adding the pins to the cache. The
-      // entries are unchanged.
-      return;
+    if (bytesToWrite > 0) {
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:418]:
```
if (bytesToWrite > 0) {
  VELOX_CHECK(!iovecs.empty());
```

File: velox/common/caching/SsdFile.cpp [hunk 10]
```diff
@@ -376,40 +383,45 @@ void SsdFile::write(std::vector<CachePin>& pins) {
...
   while (storeIndex < pins.size()) {
-    auto space = getSpace(pins, storeIndex);
+    const auto space = getSpace(pins, storeIndex);
     if (!space.has_value()) {
       // No space can be reclaimed. The pins are freed when the caller is freed.
       return;
     }
 
     auto [offset, available] = space.value();
     int32_t numWritten = 0;
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:393]:
```
    int32_t numWrittenEntries = 0;
    uint64_t writeOffset = offset;
    int32_t writeLength = 0;
    std::vector<iovec> writeIovecs;
```

File: velox/common/caching/SsdFile.cpp [hunk 11]
```diff
@@ -376,40 +383,45 @@ void SsdFile::write(std::vector<CachePin>& pins) {
 
   int32_t storeIndex = 0;
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:384]:
```
nit: s/storeIndex/writeIndex/
```

**Review by xiaoxmeng — COMMENTED** [review]
@zacw7 thanks for the update % minors

File: velox/common/caching/SsdFile.cpp [hunk 12]
```diff
@@ -438,6 +452,25 @@ void SsdFile::write(std::vector<CachePin>& pins) {
...
+    const std::vector<iovec>& iovecs) {
+  const auto ret = folly::pwritev(fd_, iovecs.data(), iovecs.size(), offset);
+  if (ret != length) {
+    VELOX_SSD_CACHE_LOG(ERROR)
+        << "Failed to write to SSD, file name: " << fileName_ << ", fd: " << fd_
+        << ", size: " << iovecs.size() << ", offset: " << offset
+        << ", error code: " << errno
+        << ", error string: " << folly::errnoStr(errno);
+    ++stats_.writeSsdErrors;
+    // If write fails, we return without adding the pins to the cache. The
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:467]: Let's put this log in the caller which has this context?

File: velox/common/caching/SsdFile.cpp [hunk 13]
```diff
@@ -438,6 +452,25 @@ void SsdFile::write(std::vector<CachePin>& pins) {
   }
 }
 
+bool SsdFile::write(
+    uint64_t offset,
+    uint64_t length,
+    const std::vector<iovec>& iovecs) {
+  const auto ret = folly::pwritev(fd_, iovecs.data(), iovecs.size(), offset);
+  if (ret != length) {
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:460]:
```
if (ret == length) {
  return true;
}
...
return false;
```

File: velox/common/caching/SsdFile.cpp [hunk 14]
```diff
@@ -374,46 +382,52 @@ void SsdFile::write(std::vector<CachePin>& pins) {
...
+      addEntryToIovecs(*entry, writeIovecs);
+      writeLength += entrySize;
+      ++numWrittenEntries;
+    }
+    if (writeLength > 0) {
+      VELOX_CHECK(!writeIovecs.empty());
+      if (!write(writeOffset, writeLength, writeIovecs)) {
+        return;
+      }
+      writeOffset += writeLength;
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.cpp:424]:
```
writeIovecs.clear();
writeLength = 0;
```

**Review by xiaoxmeng — APPROVED** [review]
@zacw7 LGTM. Please also run this patch in interactive cluster before land. Thanks!

File: velox/common/caching/SsdFile.cpp (hunk 12 above)
Comment (zacw7) [reply velox/common/caching/SsdFile.cpp:467]: which means we need to write this multiple-line code twice? if that's the case maybe there's no need to have this new write function?

**Pull request closed without merging**